compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs

```diff
@@ -1720,9 +1722,10 @@ convertExpr env0 e = do
     go env (VarIn GHC_Types "I#") args = pure (mkIdentity TInt64, args)
         -- we pretend Int and Int# are the same thing
 
-    go env (VarIn DA_Internal_Desugar "codeGenAllowLargeTuples") (LType _ : LExpr head : rest)
-        = let env' = env { envAllowLargeTuples = AllowLargeTuples True }
-           in go env' head rest
+    go env (VarIn GHC_Types "magic") (LType (isStrLitTy -> Just "userWrittenTuple") : LType _ : LExpr head : rest)
```

Comment:
looking at this I realized I'd forgotten how we actually use `magic` - it's not like `primitive` where we use the Symbol tyarg to determine what to do, rather we just use it as a placeholder in definitions like `foo = magic @"..."`, and it's the name (and module) that we use to determine the translation to LF. 

This made me wonder, couldn't we instead define `userWrittenTuple : forall a. a -> a; userWrittenTuple = magic @"userWrittenTuple` in a `daml-prim` module and change this line to

```suggestion
    go env (VarIn SOME_MODULE "userWrittenTuple") (LType _ : LExpr head : rest)
```

?

`SOME_MODULE` couldn't be `GHC.Types` since that's a stable package module, but we could just add a new non-stable package module in `daml-prim`, e.g. `GHC.Tuple.Check`

Reply:
I've implemented this in https://github.com/digital-asset/daml/pull/15445, since it constitutes a lot of changes I'd like someone to give it a look once over